## Analysing atp tennis matches with pandas

The dataset we will manipulate with consists of men's tennis matches at Wimbledon.<br><br>Many different approaches are possible so what I am trying to do here is to draw the attention to points that might not seem to be obvious at the first glance.<br>The data I am using can be downloaded in .csv format from the following link.<br>

## 1. Show installed versions

<br><br>https://raw.githubusercontent.com/solajozsef/ipython-notebooks/main/AtpWimbledonMatches.csv<br><br>If you need more data it is available at<br><br>http://tennis-data.co.uk<br><br>and there is a 'notes.txt' page explaining the abbreviations on <br><br>http://tennis-data.co.uk/notes.txt<br>

### Importing some libraries

In [44]:
import numpy as np
import pandas as pd
import math
import random
import glob

<br>This code provides multiple outputs from one code cell.<br>

In [48]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

<br>The key player of such a notebook is python pandas so let's check which version we use.<br><br>

In [3]:
!pip show pandas

Name: pandas
Version: 1.2.3
Summary: Powerful data structures for data analysis, time series, and statistics
Home-page: https://pandas.pydata.org
Author: None
Author-email: None
License: BSD
Location: /home/sj/.local/lib/python3.8/site-packages
Requires: numpy, python-dateutil, pytz
Required-by: visions, seaborn, phik, pandas-profiling


Settting some display options.

In [4]:
pd.set_option("display.max_rows", 3000, "display.max_columns", 50)
pd.get_option("display.max_rows"),
pd.get_option("display.max_columns")

(3000,)

50

<br>The dataset we are using can be downloaded from the following github link.<br><br>

In [7]:
url = 'https://raw.githubusercontent.com/solajozsef/ipython-notebooks/main/AtpWimbledonMatches.csv'

<br>Because it is a .csv file we have to read it in a pandas dataframe.<br><br>

In [8]:
df = pd.read_csv(url)

<br>Some basic info about df columns.<br><br>

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1905 entries, 0 to 1904
Data columns (total 21 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Location    1905 non-null   object 
 1   Tournament  1905 non-null   object 
 2   Date        1905 non-null   object 
 3   Winner      1905 non-null   object 
 4   Loser       1905 non-null   object 
 5   WRank       1905 non-null   int64  
 6   LRank       1905 non-null   int64  
 7   W1          1898 non-null   float64
 8   L1          1898 non-null   float64
 9   W2          1885 non-null   float64
 10  L2          1886 non-null   float64
 11  W3          1858 non-null   float64
 12  L3          1858 non-null   float64
 13  W4          922 non-null    float64
 14  L4          922 non-null    float64
 15  W5          358 non-null    float64
 16  L5          358 non-null    float64
 17  Wsets       1899 non-null   float64
 18  Lsets       1899 non-null   float64
 19  B365W       1897 non-null  

In [10]:
df.shape

(1905, 21)

<br>As we see there are 1905 rows and 17 columns in the dataframe we are going to play with.<br><br>

In [11]:
df.describe()

WRank        LRank           W1           L1           W2  \
count  1905.000000  1905.000000  1898.000000  1898.000000  1885.000000   
mean     49.756955    88.554331     5.800843     4.278714     5.855703   
std      70.233601    95.966322     1.196269     1.815351     1.184819   
min       1.000000     1.000000     0.000000     0.000000     0.000000   
25%      11.000000    36.000000     6.000000     3.000000     6.000000   
50%      30.000000    70.000000     6.000000     4.000000     6.000000   
75%      68.000000   109.000000     6.000000     6.000000     6.000000   
max    1065.000000  1085.000000     7.000000     7.000000     7.000000   

                L2           W3           L3          W4          L4  \
count  1886.000000  1858.000000  1858.000000  922.000000  922.000000   
mean      4.229056     5.824004     4.043595    5.805857    4.053145   
std       1.792565     1.142381     1.835665    1.237422    1.810932   
min       0.000000     1.000000     0.000000    0.000000    0.000000   
25%       3.000000     6.000000     3.000000    6.000000    3.000000   
50%       4.000000     6.000000     4.000000    6.000000    4.000000   
75%       6.000000     6.000000     6.000000    6.000000    6.000000   
max       7.000000     7.000000     7.000000    7.000000    7.000000   

               W5          L5        Wsets        Lsets        B365W  \
count  358.000000  358.000000  1899.000000  1899.000000  1897.000000   
mean     7.178771    4.444134     2.938915     0.678252     1.777132   
std      4.082675    4.441459     0.344239     0.772421     1.696712   
min      2.000000    0.000000     0.000000     0.000000     1.000000   
25%      6.000000    3.000000     3.000000     0.000000     1.110000   
50%      6.000000    4.000000     3.000000     0.000000     1.300000   
75%      7.000000    5.000000     3.000000     1.000000     1.720000   
max     70.000000   68.000000     3.000000     2.000000    29.000000   

             B365L  
count  1899.000000  
mean      5.196707  
std       5.187248  
min       1.002000  
25%       2.000000  
50%       3.400000  
75%       6.500000  
max      34.000000

<br>We can check the first 5 rows of df.<br><br>

In [12]:
df.head()

Location Tournament      Date         Winner             Loser  WRank  \
0   London  Wimbledon  29/06/15      Broady L.      Matosevic M.    182   
1   London  Wimbledon  29/06/15       Cilic M.         Moriya H.      9   
2   London  Wimbledon  29/06/15    Berankis R.  Haider-Maurer A.     90   
3   London  Wimbledon  29/06/15       Thiem D.           Sela D.     30   
4   London  Wimbledon  29/06/15  Granollers M.     Tipsarevic J.     72   

   LRank   W1   L1   W2   L2   W3   L3   W4   L4   W5   L5  Wsets  Lsets  \
0    138  5.0  7.0  4.0  6.0  6.0  3.0  6.0  2.0  6.0  3.0    3.0    2.0   
1    174  6.0  3.0  6.0  2.0  7.0  6.0  NaN  NaN  NaN  NaN    3.0    0.0   
2     57  6.0  2.0  5.0  2.0  NaN  NaN  NaN  NaN  NaN  NaN    1.0    0.0   
3     85  2.0  6.0  6.0  3.0  6.0  4.0  6.0  4.0  NaN  NaN    3.0    1.0   
4    486  6.0  3.0  6.0  4.0  6.0  2.0  NaN  NaN  NaN  NaN    3.0    0.0   

   B365W  B365L  
0   2.62   1.44  
1   1.02  17.00  
2   1.50   2.50  
3   1.53   2.37  
4   1.57   2.25

<br>Filling NaN(Not a Number) values with 0 and check the first 5 rows.<br><br>

In [27]:
df.fillna(value=0, inplace=True)

In [28]:
df.head()

location tournament      date         winner             loser  wrank  \
0   London  Wimbledon  29/06/15      broady l.      matosevic m.    182   
1   London  Wimbledon  29/06/15       cilic m.         moriya h.      9   
2   London  Wimbledon  29/06/15    berankis r.  haider-maurer a.     90   
3   London  Wimbledon  29/06/15       thiem d.           sela d.     30   
4   London  Wimbledon  29/06/15  granollers m.     tipsarevic j.     72   

   lrank   w1   l1   w2   l2   w3   l3   w4   l4   w5   l5  wsets  lsets  \
0    138  5.0  7.0  4.0  6.0  6.0  3.0  6.0  2.0  6.0  3.0    3.0    2.0   
1    174  6.0  3.0  6.0  2.0  7.0  6.0  0.0  0.0  0.0  0.0    3.0    0.0   
2     57  6.0  2.0  5.0  2.0  0.0  0.0  0.0  0.0  0.0  0.0    1.0    0.0   
3     85  2.0  6.0  6.0  3.0  6.0  4.0  6.0  4.0  0.0  0.0    3.0    1.0   
4    486  6.0  3.0  6.0  4.0  6.0  2.0  0.0  0.0  0.0  0.0    3.0    0.0   

   b365w  b365l  
0   2.62   1.44  
1   1.02  17.00  
2   1.50   2.50  
3   1.53   2.37  
4   1.57   2.25

<br>We can collect df's column names with<br><br>

In [19]:
df.columns

Index(['Location', 'Tournament', 'Date', 'Winner', 'Loser', 'WRank', 'LRank',
       'W1', 'L1', 'W2', 'L2', 'W3', 'L3', 'W4', 'L4', 'W5', 'L5', 'Wsets',
       'Lsets', 'B365W', 'B365L'],
      dtype='object')

<br>I like changing capital letters in columns' names, makes life easier.<br><br>

In [20]:
df.columns = df.columns.str.lower()

In [21]:
df.columns

Index(['location', 'tournament', 'date', 'winner', 'loser', 'wrank', 'lrank',
       'w1', 'l1', 'w2', 'l2', 'w3', 'l3', 'w4', 'l4', 'w5', 'l5', 'wsets',
       'lsets', 'b365w', 'b365l'],
      dtype='object')

<br>I also recommend changing 'winner' and 'loser' columns' contents to lowercase because we are going to use them a lot.<br><br>

In [22]:
df['winner'] = df['winner'].str.lower()

In [23]:
df['loser'] = df['loser'].str.lower()

<br>and check if it worked<br><br>

In [29]:
df.head(10)

location tournament      date         winner             loser  wrank  \
0   London  Wimbledon  29/06/15      broady l.      matosevic m.    182   
1   London  Wimbledon  29/06/15       cilic m.         moriya h.      9   
2   London  Wimbledon  29/06/15    berankis r.  haider-maurer a.     90   
3   London  Wimbledon  29/06/15       thiem d.           sela d.     30   
4   London  Wimbledon  29/06/15  granollers m.     tipsarevic j.     72   
5   London  Wimbledon  29/06/15      goffin d.       zeballos h.     15   
6   London  Wimbledon  29/06/15    verdasco f.         klizan m.     43   
7   London  Wimbledon  29/06/15     kyrgios n.    schwartzman d.     29   
8   London  Wimbledon  29/06/15       mayer l.     kokkinakis t.     21   
9   London  Wimbledon  29/06/15       isner j.          soeda g.     17   

   lrank   w1   l1   w2   l2   w3   l3   w4   l4    w5    l5  wsets  lsets  \
0    138  5.0  7.0  4.0  6.0  6.0  3.0  6.0  2.0   6.0   3.0    3.0    2.0   
1    174  6.0  3.0  6.0  2.0  7.0  6.0  0.0  0.0   0.0   0.0    3.0    0.0   
2     57  6.0  2.0  5.0  2.0  0.0  0.0  0.0  0.0   0.0   0.0    1.0    0.0   
3     85  2.0  6.0  6.0  3.0  6.0  4.0  6.0  4.0   0.0   0.0    3.0    1.0   
4    486  6.0  3.0  6.0  4.0  6.0  2.0  0.0  0.0   0.0   0.0    3.0    0.0   
5    128  7.0  6.0  6.0  1.0  6.0  1.0  0.0  0.0   0.0   0.0    3.0    0.0   
6     39  4.0  6.0  6.0  2.0  6.0  3.0  6.0  7.0  13.0  11.0    3.0    2.0   
7     64  6.0  0.0  6.0  2.0  7.0  6.0  0.0  0.0   0.0   0.0    3.0    0.0   
8     71  7.0  6.0  7.0  6.0  6.0  4.0  0.0  0.0   0.0   0.0    3.0    0.0   
9     91  7.0  6.0  6.0  4.0  6.0  4.0  0.0  0.0   0.0   0.0    3.0    0.0   

   b365w  b365l  
0   2.62   1.44  
1   1.02  17.00  
2   1.50   2.50  
3   1.53   2.37  
4   1.57   2.25  
5   1.11   6.50  
6   1.28   3.50  
7   1.04  13.00  
8   1.53   2.37  
9   1.06  10.00

<br>Further checking the date column we could notice it uses varying date formats so to make them uniform run the following one-liner.<br><br>

In [32]:
df.date = pd.to_datetime(df.date)

<br>Check the modified look.<br><br>

In [33]:
df.head()

location tournament       date         winner             loser  wrank  \
0   London  Wimbledon 2015-06-29      broady l.      matosevic m.    182   
1   London  Wimbledon 2015-06-29       cilic m.         moriya h.      9   
2   London  Wimbledon 2015-06-29    berankis r.  haider-maurer a.     90   
3   London  Wimbledon 2015-06-29       thiem d.           sela d.     30   
4   London  Wimbledon 2015-06-29  granollers m.     tipsarevic j.     72   

   lrank   w1   l1   w2   l2   w3   l3   w4   l4   w5   l5  wsets  lsets  \
0    138  5.0  7.0  4.0  6.0  6.0  3.0  6.0  2.0  6.0  3.0    3.0    2.0   
1    174  6.0  3.0  6.0  2.0  7.0  6.0  0.0  0.0  0.0  0.0    3.0    0.0   
2     57  6.0  2.0  5.0  2.0  0.0  0.0  0.0  0.0  0.0  0.0    1.0    0.0   
3     85  2.0  6.0  6.0  3.0  6.0  4.0  6.0  4.0  0.0  0.0    3.0    1.0   
4    486  6.0  3.0  6.0  4.0  6.0  2.0  0.0  0.0  0.0  0.0    3.0    0.0   

   b365w  b365l  
0   2.62   1.44  
1   1.02  17.00  
2   1.50   2.50  
3   1.53   2.37  
4   1.57   2.25

<br>We can sort our dataframe by date for example.<br><br>

In [38]:
df.sort_values(by='date', ascending=True, ignore_index=True).head()

location tournament       date            winner         loser  wrank  \
0   London  Wimbledon 2005-01-07        federer r.     hewitt l.      1   
1   London  Wimbledon 2005-02-07        roddick a.  johansson t.      4   
2   London  Wimbledon 2005-03-07        federer r.    roddick a.      1   
3   London  Wimbledon 2005-06-20          mayer f.    ventura s.     57   
4   London  Wimbledon 2005-06-20  philippoussis m.       beck k.    142   

   lrank   w1   l1   w2   l2   w3   l3   w4   l4   w5   l5  wsets  lsets  \
0      2  6.0  3.0  6.0  4.0  7.0  6.0  0.0  0.0  0.0  0.0    3.0    0.0   
1     22  6.0  7.0  6.0  2.0  7.0  6.0  7.0  6.0  0.0  0.0    3.0    1.0   
2      4  6.0  2.0  7.0  6.0  6.0  4.0  0.0  0.0  0.0  0.0    3.0    0.0   
3    110  6.0  4.0  7.0  5.0  6.0  2.0  0.0  0.0  0.0  0.0    3.0    0.0   
4     43  7.0  5.0  6.0  4.0  6.0  2.0  0.0  0.0  0.0  0.0    3.0    0.0   

   b365w  b365l  
0   1.11    6.5  
1   1.19    4.5  
2   1.14    5.5  
3   1.06    8.0  
4   1.66    2.1

In [36]:
df.sort_values(by=['winner', 'loser'], ascending=True, ignore_index=True).iloc[:50, :]

location tournament       date       winner              loser  wrank  \
0    London  Wimbledon 2006-06-27    agassi a.       pashanski b.     20   
1    London  Wimbledon 2006-06-29    agassi a.           seppi a.     20   
2    London  Wimbledon 2017-04-07     albot r.          bagnis f.    108   
3    London  Wimbledon 2018-04-07     albot r.          bedene a.     98   
4    London  Wimbledon 2018-03-07     albot r.   carreno busta p.     98   
5    London  Wimbledon 2016-06-29     albot r.           elias g.    110   
6    London  Wimbledon 2009-06-24   almagro n.            beck k.     48   
7    London  Wimbledon 2016-06-27   almagro n.     dutra silva r.     47   
8    London  Wimbledon 2008-06-24   almagro n.      granollers m.     12   
9    London  Wimbledon 2011-06-23   almagro n.           isner j.     15   
10   London  Wimbledon 2009-06-22   almagro n.          monaco j.     48   
11   London  Wimbledon 2011-06-21   almagro n.        nieminen j.     15   
12   London  Wimbledon 2012-06-26   almagro n.          rochus o.     11   
13   London  Wimbledon 2013-06-26   almagro n.           rufin g.     16   
14   London  Wimbledon 2012-06-27   almagro n.           rufin g.     11   
15   London  Wimbledon 2013-06-24   almagro n.            zopp j.     16   
16   London  Wimbledon 2009-06-23     alves t.           pavel a.    118   
17   London  Wimbledon 2006-06-27     ancic m.         almagro n.     10   
18   London  Wimbledon 2006-03-07     ancic m.        djokovic n.     10   
19   London  Wimbledon 2008-06-27     ancic m.          ferrer d.     43   
20   London  Wimbledon 2008-06-23     ancic m.          llodra m.     43   
21   London  Wimbledon 2005-06-24     ancic m.         monfils g.     17   
22   London  Wimbledon 2008-06-25     ancic m.      petzschner p.     43   
23   London  Wimbledon 2006-06-28     ancic m.     sanguinetti d.     10   
24   London  Wimbledon 2005-06-20     ancic m.        summerer t.     17   
25   London  Wimbledon 2005-06-22     ancic m.       udomchoke d.     17   
26   London  Wimbledon 2008-06-30     ancic m.        verdasco f.     43   
27   London  Wimbledon 2006-01-07     ancic m.        wawrinka s.     10   
28   London  Wimbledon 2014-06-23  anderson k.          bedene a.     18   
29   London  Wimbledon 2017-07-07  anderson k.       bemelmans r.     42   
30   London  Wimbledon 2018-11-07  anderson k.         federer r.      8   
31   London  Wimbledon 2014-06-27  anderson k.         fognini f.     18   
32   London  Wimbledon 2018-02-07  anderson k.          gombos n.      8   
33   London  Wimbledon 2019-01-07  anderson k.       herbert p.h.      8   
34   London  Wimbledon 2015-01-07  anderson k.           ilhan m.     14   
35   London  Wimbledon 2018-07-13  anderson k.           isner j.      8   
36   London  Wimbledon 2018-06-07  anderson k.   kohlschreiber p.      8   
37   London  Wimbledon 2011-06-22  anderson k.       marchenko i.     36   
38   London  Wimbledon 2015-03-07  anderson k.           mayer l.     14   
39   London  Wimbledon 2018-09-07  anderson k.         monfils g.      8   
40   London  Wimbledon 2015-06-29  anderson k.         pouille l.     14   
41   London  Wimbledon 2013-06-27  anderson k.      przysiezny m.     23   
42   London  Wimbledon 2013-06-25  anderson k.          rochus o.     23   
43   London  Wimbledon 2014-06-25  anderson k.  roger-vasselin e.     18   
44   London  Wimbledon 2017-05-07  anderson k.           seppi a.     42   
45   London  Wimbledon 2018-05-07  anderson k.           seppi a.      8   
46   London  Wimbledon 2019-03-07  anderson k.      tipsarevic j.      8   
47   London  Wimbledon 2017-03-07  anderson k.        verdasco f.     42   
48   London  Wimbledon 2011-06-21   andreev i.      gabashvili t.     79   
49   London  Wimbledon 2012-06-25   andreev i.         golding o.     93   

    lrank   w1   l1   w2   l2   w3   l3   w4   l4    w5    l5  wsets  lsets  \
0      71  2.0  6.0  6.0  2.0  6.0  4.0  6.0  3

So let's get to it.


Let's see how many favorite players with odds lower than say 1.30 won the match and how many lost. The winners odds are in the 'b365w' column.

To get a clear view we need to know how many players with odds lower than 1.3 are in the winners' column(b365w) and how many are in the losers' column(b365l).

In [14]:
df.query('(b365w <= 1.3) or (b365l <= 1.3) ').b365w.count()

1104

In [15]:
df.query('(b365w <= 1.3)').b365w.count()

961

<br>When the favorite loses his odds appear in the 'b365l' column.<br><br>

In [16]:
df.query('(b365l <= 1.3)').b365l.count()

143

<br>What we see here is that out of 1104 matches the favorites won 961 matches and lost 143 ones.<br><br>

<br>How many winners' ranks are above say 900 in world ranking.<br><br>

In [17]:
df[df['wrank'] > 900].count().b365w

2

<br>and who are those...<br><br>

In [18]:
df[df['wrank'] > 900]

location tournament      date     winner        loser  wrank  lrank   w1  \
1320   London  Wimbledon  26/06/07  Kiefer N.  Volandri F.   1065     27  6.0   
1354   London  Wimbledon  28/06/07  Kiefer N.   Santoro F.   1065     70  6.0   

       l1   w2   l2   w3   l3  wsets  lsets  b365w  b365l  
1320  3.0  7.0  6.0  6.0  1.0    3.0    0.0   1.16   4.50  
1354  4.0  6.0  3.0  6.0  4.0    3.0    0.0   2.37   1.53

<br>Say I would like to see the favorite losers where the favorite odds were between 1.4 and 1.5<br><br>

In [19]:
df.query('(b365l >= 1.4) and (b365l < 1.5)').count().b365w

51

<br>... and the total number of games within these favorite odds<br><br>

In [20]:
df.query('((b365w >= 1.4) and (b365w < 1.5)) or ((b365l >= 1.4) and (b365l < 1.5))').count().b365w

150

<br>So altogether there were 150 games within these favorite odds and the favorite won 51 matches.<br><br>

<br>Underdogs with odds higher than 4.0 winning the match. <br><br>

In [18]:
df.query('b365w >= 4').count().b365w

109

<br>And underdogs with odds higher than 4.0 losing the match. <br><br>

In [21]:
df.query('b365l >= 4').count().b365w

845

<br>Let's see how tennis aces perfom against each other. The following cell filters those games where players' world rankings were between 1 and 10 and we printed the relevant cells.<br><br>wrank: winner's rank<br>lrank: loser's rank<br>b365w: winner's odds<br>b365l: loser's odds<br><br>The first column is the row number column in our table.<br>

In [22]:
df.query('((wrank >= 1) and (wrank <= 10 )) and ((lrank >= 1) and (lrank <= 10 ) and (b365w < b365l))')[['wrank', 'lrank', 'b365w', 'b365l']],

(      wrank  lrank  b365w  b365l
 123       1      9   1.08   7.50
 126       1      2   1.83   2.00
 247       1     10   1.05   9.00
 253       1      2   1.16   5.00
 380       2      6   1.11   6.50
 505       1      2   1.11   6.50
 507       1      4   1.14   5.50
 630       1      6   1.66   2.20
 631       4     10   1.22   4.33
 633       1      4   1.57   2.37
 757       3      7   1.11   6.50
 761       1      3   1.55   2.60
 884       5      7   1.20   4.50
 888       5      6   1.22   4.50
 1011      1      9   1.14   5.50
 1014      1      4   1.50   2.62
 1136      1      6   1.10   7.00
 1140      1      8   1.16   5.00
 1268      2      9   1.10   7.00
 1269      2      7   1.25   4.33
 1382      1     10   1.01  15.00
 1395      2      5   1.25   4.00
 1396      1      2   1.14   5.50
 1519      4      3   1.28   3.75
 1522      4      9   1.40   3.00
 1523      2      4   1.61   2.40
 1646      1      4   1.36   3.20
 1900      4      5   1.44   2.75
 1903      4  

In [23]:
df.query('((wrank >= 1) and (wrank <= 10 )) and ((lrank >= 1) and (lrank <= 10 ) and (b365l < b365w))')[['wrank', 'lrank', 'b365w', 'b365l']],

(      wrank  lrank  b365w  b365l
 379       6      3   3.50   1.30
 760       3      2   2.10   1.72
 1005      9      7   2.87   1.40
 1015      2      1   2.30   1.66
 1137      8      4   2.10   1.72
 1142      2      1   2.50   1.57
 1267      7      3   2.62   1.50
 1645      8      2  10.00   1.07
 1648      8     10   2.00   1.80
 1777      2      1   2.20   1.66
 1897      5      9   2.20   1.66
 1902      3      1   2.87   1.44,)

<br>The first table shows the matches where the favorite won and the second where the favorite lost.<br>

Let's focus on very strong favorites whose booky odds are below 1.10 and see how they perform.<br>First we check how many matches they won.<br><br>

In [24]:
df.query('b365w <= 1.1').count().b365w

458

<br>Now how many matches they lost<br><br>

In [25]:
df.query('b365l <= 1.1').count().b365w

24

<br>The figures are quite impressive. We can say that strong favorites with odds lower than 1.10 win the majority of the games. <br><br>

In [26]:
df.query('b365l <= 1.1').head()

location tournament      date            winner          loser  wrank  \
29    London  Wimbledon  29/06/15          Ilhan M.    Janowicz J.     82   
46    London  Wimbledon  30/06/15  Ramos-Vinolas A.     Istomin D.     65   
223   London  Wimbledon  30/06/06       Verdasco F.  Nalbandian D.     30   
317   London  Wimbledon  24/06/09  Gimeno-Traver D.        Dent T.     98   
464   London  Wimbledon  23/06/05       Tursunov D.      Henman T.    152   

     lrank   w1   l1   w2   l2   w3   l3  wsets  lsets  b365w  b365l  
29      47  7.0  6.0  6.0  4.0  6.0  7.0    3.0    1.0    7.0   1.10  
46      62  6.0  2.0  6.0  2.0  3.0  2.0    2.0    0.0    7.0   1.10  
223      3  7.0  6.0  7.0  6.0  6.0  2.0    3.0    0.0    7.0   1.08  
317    266  7.0  5.0  7.0  6.0  4.0  6.0    3.0    2.0    7.0   1.08  
464      9  3.0  6.0  6.0  2.0  3.0  6.0    3.0    2.0    7.0   1.08

<br>We can search for players by name like so<br><br>

In [27]:
name = 'Nadal'
wn = df.winner.str.contains(name)
ls = df.loser.str.contains(name)

<br>Above we created two variables 'wn' and 'ln'. 'wn' stands for winner's name and 'ln' for loser's name.<br>If we want to see how many times this player won or lost, the code is below.<br>Note: if we want to refer to a variable within a query we use the '@' character<br><br>

In [28]:
print('win: ', df.query('@wn == True').shape[0],  'times')
print('lose: ', df.query('@ls == True').shape[0], 'times')

win:  51 times
lose:  11 times


<br>The same process with Thiem.<br><br>

In [29]:
name = 'Thiem'
wn = df.winner.str.contains(name)
ls = df.loser.str.contains(name)

In [30]:
print('win: ', df.query('@wn == True').shape[0],  'times')
print('lose: ', df.query('@ls == True').shape[0], 'times')

win:  5 times
lose:  6 times


<br>We can even write a little function for this operation.<br><br>

In [31]:
def win_lose(name):
    wn = df.winner.str.contains(name)
    ls = df.loser.str.contains(name)
    
    if df.winner.str.contains(name).any():
      #  var = df.query('@mm == True').shape[0]
        print('win:  ', df.query('@wn == True').shape[0],  'times')
        
    if df.loser.str.contains(name).any():
       # var = df.query('@mm == True').shape[0]
        print('lose: ', df.query('@ls == True').shape[0],  'times')

<br>and then we only have to modify the 'name' variable.<br><br>

In [34]:
name = 'Cilic'
win_lose(name)

win:   29 times
lose:  13 times


In [35]:
name = 'Berankis'
win_lose(name)

win:   3 times
lose:  7 times


In [37]:
name = 'Djoko'
win_lose(name)

win:   72 times
lose:  10 times


<br>If you want to know who has defeated Djokovic one possible way is below:<br><br>

In [43]:
df[df.loser.str.contains('kovic') == True][['winner', 'loser', 'b365w', 'b365l']]

winner        loser  b365w  b365l
243      ancic m.  djokovic n.   1.36   3.00
375       haas t.  djokovic n.   3.20   1.36
486   grosjean s.  djokovic n.   1.14   5.00
632    berdych t.  djokovic n.   2.00   1.80
885    berdych t.  djokovic n.   4.00   1.25
1142    murray a.  djokovic n.   2.50   1.57
1243   querrey s.  djokovic n.  26.00   1.01
1395     nadal r.  djokovic n.   1.25   4.00
1718     safin m.  djokovic n.   9.00   1.05
1902   federer r.  djokovic n.   2.87   1.44

<br>Or how the matches ended between Djokovic and Federer. One thing to note here: pandas' contains() method returns results of partial matches too. So if you do not remember the exact name of a player only a few characters will do the trick like instead of 'Djokovic' I used ' joko' only. This code still works.<br><br>

In [42]:
df[df.loser.str.contains('joko|edere') & df.winner.str.contains('joko|edere')]

location tournament       date       winner        loser  wrank  lrank  \
126    London  Wimbledon 2015-12-07  djokovic n.   federer r.      1      2   
761    London  Wimbledon 2019-07-14  djokovic n.   federer r.      1      3   
1523   London  Wimbledon 2014-06-07  djokovic n.   federer r.      2      4   
1902   London  Wimbledon 2012-06-07   federer r.  djokovic n.      3      1   

       w1   l1   w2   l2   w3   l3   w4   l4    w5    l5  wsets  lsets  b365w  \
126   7.0  6.0  6.0  7.0  6.0  4.0  6.0  3.0   0.0   0.0    3.0    1.0   1.83   
761   7.0  6.0  1.0  6.0  7.0  6.0  4.0  6.0  13.0  12.0    3.0    2.0   1.55   
1523  6.0  7.0  6.0  4.0  7.0  6.0  5.0  7.0   6.0   4.0    3.0    2.0   1.61   
1902  6.0  3.0  3.0  6.0  6.0  4.0  6.0  3.0   0.0   0.0    3.0    1.0   2.87   

      b365l  
126    2.00  
761    2.60  
1523   2.40  
1902   1.44

<br>Another way is when we sort out the df by the 'winner' and 'loser' columns so we can see who was playing whom.Many options arise here.<br><br>

In [39]:
df.sort_values(by=['winner', 'loser'], ascending=True, ignore_index=True).iloc[1500:1650, :]

location tournament       date             winner                 loser  \
1500   London  Wimbledon 2005-06-21          rochus o.          goldstein p.   
1501   London  Wimbledon 2006-06-27          rochus o.             muller g.   
1502   London  Wimbledon 2008-06-23          rochus o.               sela d.   
1503   London  Wimbledon 2006-06-29          rochus o.                zib t.   
1504   London  Wimbledon 2005-06-25         roddick a.            andreev i.   
1505   London  Wimbledon 2012-06-27         roddick a.              baker j.   
1506   London  Wimbledon 2011-06-21         roddick a.               beck a.   
1507   London  Wimbledon 2009-06-29         roddick a.            berdych t.   
1508   London  Wimbledon 2005-06-24         roddick a.          bracciali d.   
1509   London  Wimbledon 2009-06-23         roddick a.             chardy j.   
1510   London  Wimbledon 2005-06-27         roddick a.              coria g.   
1511   London  Wimbledon 2007-06-25         roddick a.          gimelstob j.   
1512   London  Wimbledon 2005-06-29         roddick a.           grosjean s.   
1513   London  Wimbledon 2011-06-22         roddick a.            hanescu v.   
1514   London  Wimbledon 2009-01-07         roddick a.             hewitt l.   
1515   London  Wimbledon 2005-02-07         roddick a.          johansson t.   
1516   London  Wimbledon 2010-06-25         roddick a.      kohlschreiber p.   
1517   London  Wimbledon 2009-06-25         roddick a.           kunitsyn i.   
1518   London  Wimbledon 2010-06-23         roddick a.             llodra m.   
1519   London  Wimbledon 2007-04-07         roddick a.          mathieu p.h.   
1520   London  Wimbledon 2006-06-29         roddick a.              mayer f.   
1521   London  Wimbledon 2009-06-27         roddick a.             melzer j.   
1522   London  Wimbledon 2009-03-07         roddick a.             murray a.   
1523   London  Wimbledon 2012-06-28         roddick a.               phau b.   
1524   London  Wimbledon 2010-06-21         roddick a.                ram r.   
1525   London  Wimbledon 2008-06-24         roddick a.            schwank e.   
1526   London  Wimbledon 2006-06-28         roddick a.         tipsarevic j.   
1527   London  Wimbledon 2007-06-27         roddick a.          udomchoke d.   
1528   London  Wimbledon 2005-06-21         roddick a.              vanek j.   
1529   London  Wimbledon 2007-06-29         roddick a.           verdasco f.   
1530   London  Wimbledon 2007-06-26  roger-vasselin e.              baker j.   
1531   London  Wimbledon 2007-06-28  roger-vasselin e.            chela j.i.   
1532   London  Wimbledon 2016-06-27  roger-vasselin e.         gabashvili t.   
1533   London  Wimbledon 2014-06-23  roger-vasselin e.           volandri f.   
1534   London  Wimbledon 2014-06-23            rola b.            andujar p.   
1535   London  Wimbledon 2012-06-26           rosol l.              dodig i.   
1536   London  Wimbledon 2015-06-30           rosol l.             gulbis e.   
1537   London  Wimbledon 2017-03-07           rosol l.          laaksonen h.   
1538   London  Wimbledon 2012-06-28           rosol l.              nadal r.   
1539   London  Wimbledon 2014-06-24           rosol l.              paire b.   
1540   London  Wimbledon 2019-02-07          rublev a.              garin c.   
1541   London  Wimbledon 2017-04-07          rublev a.          travaglia s.   
1542   London  Wimbledon 2012-06-25           rufin g.             darcis s.   
1543   London  Wimbledon 2013-06-24           rufin g.          matosevic m.   
1544   London  Wimbledon 2005-06-20        rusedski g.             martin a.   
1545   London  Wimbledon 2012-06-25         russell m.  menendez-maceiras a.   
1546   London  Wimbledon 2010-06-22         russell m.        riba-madrid p.   
1547   London  Wimbledon 2007-06-26           safin m.           de voest r.   
1548   London  Wimbledon 2008-06-25           safin m.           djokovic n.   
1549 

<br>if you do not need that many columns...<br>

In [41]:
df.sort_values(by=['winner', 'loser'], ascending=True, ignore_index=True).iloc[600:900, :][['date', 'winner', 'b365w', 'loser', 'b365l']]

date            winner  b365w               loser  b365l
600 2010-06-24         ferrer d.   1.22            serra f.   4.00
601 2006-06-28         ferrer d.   1.39          stadler s.   2.75
602 2008-06-23         ferrer d.   1.06       stakhovsky s.   8.00
603 2007-06-29      ferrero j.c.   4.00            blake j.   1.22
604 2005-06-20      ferrero j.c.   1.28          delgado j.   3.50
605 2009-06-27      ferrero j.c.   2.62         gonzalez f.   1.44
606 2007-06-26      ferrero j.c.   1.14            hajek j.   5.00
607 2006-06-28      ferrero j.c.   1.30        karanusic r.   3.39
608 2005-06-22      ferrero j.c.   1.38            lee h.t.   2.87
609 2005-06-25      ferrero j.c.   1.66            mayer f.   2.10
610 2007-06-28      ferrero j.c.   1.50           muller g.   2.50
611 2008-06-23      ferrero j.c.   1.61          querrey s.   2.20
612 2009-06-25      ferrero j.c.   1.36          santoro f.   3.00
613 2009-06-29      ferrero j.c.   1.66            simon g.   2.10
614 2007-03-07      ferrero j.c.   1.61       tipsarevic j.   2.25
615 2006-06-27      ferrero j.c.   1.14          vicente f.   5.00
616 2009-06-23      ferrero j.c.   2.20          youzhny m.   1.61
617 2010-06-22        fischer m.   1.50            soeda g.   2.50
618 2011-06-27           fish m.   2.87          berdych t.   1.40
619 2006-06-28           fish m.   1.66          ginepri r.   2.10
620 2012-06-30           fish m.   2.25           goffin d.   1.61
621 2011-06-20           fish m.   1.10       granollers m.   6.50
622 2011-06-25           fish m.   1.28            haase r.   3.50
623 2011-06-22           fish m.   1.25          istomin d.   3.75
624 2012-06-26           fish m.   1.14  ramirez-hidalgo r.   5.50
625 2009-06-22           fish m.   1.01          roitman s.  13.00
626 2009-06-24           fish m.   1.61       tipsarevic j.   2.20
627 2010-06-21           fish m.   1.14            tomic b.   5.00
628 2006-06-29           fish m.   1.06     van gemerden m.   8.00
629 2012-06-28           fish m.   1.36             ward j.   3.00
630 2018-05-07        fognini f.   1.53          bolelli s.   2.50
631 2018-03-07        fognini f.   1.28           daniel t.   3.75
632 2016-06-30        fognini f.   1.33         delbonis f.   3.40
633 2019-04-07        fognini f.   1.66        fucsovics m.   2.20
634 2009-06-23        fognini f.   3.75          istomin d.   1.25
635 2014-06-23        fognini f.   1.22       kuznetsov al.   4.00
636 2012-06-25        fognini f.   2.62           llodra m.   1.44
637 2014-06-25        fognini f.   1.28            puetz t.   3.50
638 2010-06-24        fognini f.   1.80          russell m.   1.90
639 2015-06-30        fognini f.   1.44          smyczek t.   2.62
640 2019-02-07        fognini f.   1.72           tiafoe f.   2.10
641 2017-03-07        fognini f.   1.10         tursunov d.   7.00
642 2010-06-22        fognini f.  10.00         verdasco f.   1.04
643 2017-05-07        fognini f.   1.61           vesely j.   2.30
644 2019-02-07          fritz t.   1.40          berdych t.   3.00
645 2018-03-07          fritz t.   1.25           sonego l.   4.00
646 2019-02-07      fucsovics m.   1.80            novak d.   2.00
647 2010-06-21     gabashvili t.   1.53          delgado r.   2.37
648 2008-06-24        galvani s.   1.57            baker j.   2.25
649 2006-06-27        galvani s.   2.50            waske a.   1.50
650 2008-06-26   garcia-lopez g.   2.62          almagro n.   1.44
651 2009-06-22   garcia-lopez g.   1.08          calleri a.   7.00
652 2018-02-07   garcia-lopez g.   1.11            elias g.   7.00
653 2005-06-21   garcia-lopez g.   2.00           ferrer d.   1.72
654 2011-06-21   garcia-lopez g.   1.36          golubev a.   3.00
655 2006-06-28   garcia-lopez g.   1.90            minar i.   1.80
656 2012-06-26   garcia-lopez g.   3.00   roger-vasselin e.   1.36
657 2008-06-24   garcia-lopez g.   1.66    van der merwe i.   2.10
658 2012-06-29        gasquet r.   1.44          almagro 

<br>Modify the code blocks as you like!